parachain/pallets/eth-app/src/lib.rs

```diff
@@ -32,12 +33,25 @@ use artemis_core::{ChannelId, SingleAsset, OutboundRouter};
 mod payload;
 use payload::OutboundPayload;
 
+mod benchmarking;
+
 #[cfg(test)]
 mod mock;
 
 #[cfg(test)]
 mod tests;
 
+/// Weight functions needed for this pallet.
+pub trait WeightInfo {
```

Comment:
I think we're supposed to dynamically generate this trait definition and its implementation for `()`. I looked at two other big projects (the Polkadot relay chain, and Acala), and they are doing roughly this for each module:

1. Calling the `benchmark` command with the `--template` parameter to generate `pallets/{module}/src/weights.rs`. The generated code includes the trait definition and its implementation for `()`.
2. Calling `benchmark` again, this time without a `--template` parameter, to generate `runtime/{runtime}/src/weights/{module}.rs`

After poking around, I found these templates used by Acala and Substrate respectively.
* https://github.com/AcalaNetwork/Acala/blob/master/templates/module-weight-template.hbs
* https://github.com/paritytech/substrate/blob/master/.maintain/frame-weight-template.hbs

And an example `weights.rs` generated for an Acala module:
https://github.com/AcalaNetwork/Acala/blob/master/modules/dex/src/weights.rs

Reply:
Anway, can leave this for a future PR.

Reply:
It looks like they're using the same template as Substrate itself: https://github.com/paritytech/substrate/blob/master/.maintain/frame-weight-template.hbs

I considered doing the same, but decided against it because I don't think the Substrate approach serves us well. As you can see at https://github.com/AcalaNetwork/Acala/blob/master/modules/dex/src/weights.rs, the template generates weights as a submodule of the pallet. This obscures the fact that weights are runtime-specific. I think there might be a temptation for a dev to use the pallet weights as-is instead of regenerating them for each runtime.

I also don't think generating the `WeightInfo` trait is a great idea. The weight(s) associated with an extrinsic are closely tied to the inner workings of the extrinsic. So defining the interface, i.e. `WeightInfo`, that describes these inner workings in a separate module feels wrong. The benchmarks should fulfil the interface but not define the interface themselves.

Pedantic, misguided or reasonable? Interested in your thoughts @vgeddes

Reply:
Hey, yeah, I think that's reasonable and makes sense - I did find it odd that they were generating weights alongside the source code for the pallet, which is supposed to be runtime-agnostic. And having to run the benchmark command twice is a bit cumbersome too. I think the benchmarking framework is still a WIP while they iterate towards a holistic solution.